conda_kapsel/archiver.py

```diff
@@ -329,3 +332,207 @@ def _archive_project(project, filename):
             pass
 
     return SimpleStatus(success=True, description=("Created project archive %s" % filename), logs=logs)
+
+
+def _list_files_zip(zip_path):
+    with zipfile.ZipFile(zip_path, mode='r') as zf:
+        return sorted(zf.namelist())
+
+
+def _list_files_tar(tar_path):
+    with tarfile.open(tar_path, mode='r') as tf:
+        # we don't want links or block devices or anything weird, they could be a security problem
+        return sorted([member.name for member in tf.getmembers() if member.isreg() or member.isdir()])
+
+
+def _extract_files_zip(zip_path, src_and_dest, logs):
+    # the zipfile API has no way to extract to a filename of
+    # our choice, so we have to unpack to a temporary location,
+    # then copy those files over.
+    tmpdir = tempfile.mkdtemp()
+    try:
+        with zipfile.ZipFile(zip_path, mode='r') as zf:
+            zf.extractall(tmpdir)
+            for (src, dest) in src_and_dest:
+                logs.append("Unpacking %s to %s" % (src, dest))
+                src_path = os.path.join(tmpdir, src)
+                if os.path.isdir(src_path):
+                    makedirs_ok_if_exists(dest)
+                    shutil.copystat(src_path, dest)
+                else:
+                    makedirs_ok_if_exists(os.path.dirname(dest))
+                    shutil.copy2(src_path, dest)
+    finally:
+        try:
+            shutil.rmtree(tmpdir)
+        except (IOError, OSError):
+            pass
+
+
+def _extract_files_tar(tar_path, src_and_dest, logs):
+    with tarfile.open(tar_path, mode='r') as tf:
+        for (src, dest) in src_and_dest:
+            logs.append("Unpacking %s to %s" % (src, dest))
+            member = tf.getmember(src)
+            # we could also use tf._extract_member here, but the
+            # solution below with only the public API isn't that
+            # bad.
+            if member.isreg():
+                makedirs_ok_if_exists(os.path.dirname(dest))
+                tf.makefile(member, dest)
+            else:
+                assert member.isdir()
```

Comment:
Since there are other possible tarfile entry types, this assertion might not be true (it's input dependent). So it might be better to raise a different error here.